### ISSUE TITLE
Conv2D operator with SAME padding when Stride > kernel size showing unexpected results

### ISSUE THREAD

**Issue by Oewyn** [tensorflow/tensorflow#14601]
### System information
- **Have I written custom code -- YES, only to demonstrate the problem (source code is below)**:
- **OS Platform and Distribution (Linux Ubuntu 16.04)**:
- **TensorFlow installed from (binary (PIP))**:
- **TensorFlow version (1.4.0)**:
- **Python version (2.7.12)**: 
- **Bazel version (N/A)**:
- **GCC/Compiler version (N/A)**:
- **CUDA/cuDNN version (N/A)**:
- **GPU model and memory (N/A -- CPU only)**:
- **Exact command to reproduce (See Source Code Below)**:

### Describe the problem
There is an inconsistency between the convolution documentation on padding with 'SAME' located [here](https://www.tensorflow.org/api_guides/python/nn#Convolution) and the behavior of the tf.nn.conv2d operator. In the example below I create a 3x1 input with values [[1.0][1.1][1.2]] and a 1x1 filter of value [1.0]. I specify the stride to be 1x3x1x1 which should result in only a single element be output and the padding to be 'SAME'. From the padding calculation in the above link: 

pad_along_height:
    
    in_height ( = 3) % strides[1]( = 3) == 0 so
    pad_along_height = max(filter_height ( = 1) - strides[1] ( = 3), 0)
    pad_along_height = max(-2, 0) = 0

pad along_width:

    in_width ( = 1) % strides[2] ( = 1) == 0 so
    pad_along_width = max(filter_width( = 1) - strides[2] ( = 1), 0
    pad_along_width = max(0,0) = 0

My hypothesis is that pad_along_* is not using the max(x,0) and as a result, pad_along_height = -2. Therefore pad_top = -1 and pad_bottom = -1. If that was the case, then our input is reduced to only the middle element [1.1] which explains why the TF result of the code below is 1.1 rather than the expected 1.0 (value of first input).

If I change the padding to be VALID (no padding) then this code below gives the result of 1.0 or if i instead change the stride to 1,2,1,1 i get the expected value of 1.0 (although in this case my hypothesis proposes that pad_bottom is still -1).

### Source code / logs
    import tensorflow as tf
    import numpy as np

    i = tf.constant((np.ones(3) + np.arange(3) * 0.1).reshape(1,3,1,1), dtype=tf.float32, name='input')
    f = tf.constant(np.ones(1).reshape(1,1,1,1), dtype=tf.float32, name='filter')

    conv = tf.nn.conv2d(input=i, filter=f, strides=(1,3,1,1), padding='SAME')

    with tf.Session() as sess:
        out = sess.run(conv)
        print out

Output:
`[[[[ 1.10000002]]]]`

**Comment by facaiy** [comment]
I believe that `max(0, pad)` works well, see:

https://github.com/tensorflow/tensorflow/blob/69620e12bf3403a11a47b26006ebb22656a9f036/tensorflow/core/framework/common_shape_fns.cc#L42-L44

And I find the stride information isn't packed into `args` in `LaunchDeepConvOp`, perhaps the problem is here but I am still not sure about it:

https://github.com/tensorflow/tensorflow/blob/69620e12bf3403a11a47b26006ebb22656a9f036/tensorflow/core/kernels/conv_ops.cc#L158-L169

**Comment by reedwm** [comment]
This only occurs on the CPU, not the GPU. The fact the CPU and GPU results differ is a serious issue.

/CC @yzhwang @rmlarsen

**Comment by yzhwang** [comment]
@Oewyn Thanks for reporting the issue.
I believe the issue is due to code here:
https://bitbucket.org/eigen/eigen/src/b6e6d0cf6a77518580b5c7070beda7101118067c/unsupported/Eigen/CXX11/src/Tensor/TensorImagePatch.h?at=default&fileviewer=file-view-default#TensorImagePatch.h-266
And you are right on how to fix this too. We have noticed this internally too and have fixed the internal Eigen by adding the max(0, x), but since the OSS TensorFlow depends on the OSS Eigen which is hosted on bitbucket, I haven't yet create a pull request to make my change there. I will try to do so today.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
A member of the TensorFlow organization has replied after the stat:awaiting tensorflower label was applied.

**Comment by reedwm** [comment]
@yzhwang this appears to be merged. Is the issue fixed?

**Comment by yzhwang** [comment]
@gunan A change to eigen that should fix this was just merged:
https://bitbucket.org/eigen/eigen/commits/2355b229ea4c
I'm not sure how do we update our eigen dependency. If it's not scheduled or done automatically, would you please update it to the latest version of eigen so that @Oewyn can verify if this will fix the issue? Thank you!

**Comment by yongtang** [comment]
@yzhwang @gunan I created a PR #16705 to update the Eigen library to `2355b229ea4c`. Please take a look.

**Comment by yzhwang** [comment]
Thanks @gunan and @yongtang !
